src/autopas/containers/linkedCells/traversals/C04Traversal.h

```diff
@@ -0,0 +1,233 @@
+/**
+ * @file C04Traversal.h
+ * @author C.Menges, based on tchipevn (original source:
+ * ls1-mardyn/src/particleContainer/LinkedCellTraversals/C04CellPairTraversal.h)
+ * @date 15.06.2019
+ */
+
+#pragma once
+
+#include "C08CellHandler.h"
+#include "LinkedCellTraversalInterface.h"
+#include "autopas/containers/cellPairTraversals/C08BasedTraversal.h"
+#include "autopas/pairwiseFunctors/CellFunctor.h"
+#include "autopas/utils/ThreeDimensionalMapping.h"
+#include "autopas/utils/WrapOpenMP.h"
+
+namespace autopas {
+
+/**
+ * This class provides the c04 traversal.
+ *
+ * The traversal uses the c04 base step performed on every single cell. Since
+ * these steps overlap a domain coloring with eight colors is applied.
+ *
+ * @tparam ParticleCell the type of cells
+ * @tparam PairwiseFunctor The functor that defines the interaction of two particles.
+ * @tparam useSoA
+ * @tparam useNewton3
+ */
+template <class ParticleCell, class PairwiseFunctor, DataLayoutOption DataLayout, bool useNewton3>
+class C04Traversal : public C08BasedTraversal<ParticleCell, PairwiseFunctor, DataLayout, useNewton3>,
+                     public LinkedCellTraversalInterface<ParticleCell> {
+ public:
+  /**
+   * Constructor of the c04 traversal.
+   * @param dims The dimensions of the cellblock, i.e. the number of cells in x,
+   * y and z direction.
+   * @param pairwiseFunctor The functor that defines the interaction of two particles.
+   * @param cutoff Cutoff radius.
+   * @param cellLength cell length.
+   */
+  C04Traversal(const std::array<unsigned long, 3> &dims, PairwiseFunctor *pairwiseFunctor, const double cutoff = 1.0,
+               const std::array<double, 3> &cellLength = {1.0, 1.0, 1.0})
+      : C08BasedTraversal<ParticleCell, PairwiseFunctor, DataLayout, useNewton3>(dims, pairwiseFunctor, cutoff,
+                                                                                 cellLength),
+        _cellHandler(pairwiseFunctor, this->_cellsPerDimension, cutoff, cellLength, this->_overlap) {
+    computeOffsets32Pack();
+  }
+
+  /**
+   * @copydoc LinkedCellTraversalInterface::traverseCellPairs()
+   */
+  void traverseCellPairs(std::vector<ParticleCell> &cells) override;
+
+  TraversalOption getTraversalType() const override { return TraversalOption::c04; }
+
+  /**
+   * C04 traversals are always usable.
+   * @return
+   */
+  bool isApplicable() const override {
```

Comment:
would probably make sense to extend the if/def around everything but the `return true`. Can we make this (base-)function then constexpr?